Sources/Fuzzilli/Corpus/MarkovCorpus.swift

```diff
@@ -0,0 +1,206 @@
+    // This allows it to build an initial baseline by randomly selecting a program to mutate
+    // before switching to the more computationally expensive selection of programs that
+    // hit infreqent edges
+    private var totalExecs: UInt32 = 0
+
+    // This scheduler returns one base program multiple times, in order to compensate the overhead caused by tracking
+    // edge counts
+    private var currentProg: Program
+    private var remainingEnergy: UInt32 = 0
+
+    // Markov corpus requires an evaluator that tracks edge coverage
+    // Thus, the corpus object keeps a reference to the evaluator, in order to only downcast once
+    private var covEvaluator: ProgramCoverageEvaluator
+
+    // Rate at which selected samples will be included, to promote diversity between instances
+    // Equivalent to 1 - dropoutRate
+    private var dropoutRate: Double
+
+    // The scheduler will initially selectd the 1 / desiredSelectionProportion samples with the least frequent
+    // edge hits in each round, before dropout is applied
+    private let desiredSelectionProportion = 8
+
+    public init(covEvaluator: ProgramCoverageEvaluator, dropoutRate: Double) {
+        self.dropoutRate = dropoutRate
+        covEvaluator.enableEdgeTracking()
+        self.covEvaluator = covEvaluator
+        self.currentProg = Program()
+        super.init(name: "MarkovCorpus")
+    }
+
+    override func initialize() {
+        assert(covEvaluator === fuzzer.evaluator as! ProgramCoverageEvaluator)
+        currentProg = makeSeedProgram()
+        add(currentProg)
+    }
+
+    // This should only be called when ProgramAspects are not available, such as on an import from
+    // another fuzzer instance
+    public func add(_ program: Program) {
+        guard program.size > 0 else { return }
+        if let aspects = execAndEvalProg(program) {
+            add(program, aspects)
+        }
+    }
+
+    public func add(_ program: Program, _ aspects: ProgramAspects) {
+        guard program.size > 0 else { return }
+
+        guard let origCov = aspects as? CovEdgeSet else {
+            logger.fatal("Markov Corpus needs to be provided a CovEdgeSet when adding a program")
+        }
+
+        covEvaluator.clearEdgeList(origCov.toEdges())
+        // One reason why this could fail is that no new edges are found the second round
+        guard let secondCov = execAndEvalProg(program) as? CovEdgeSet else {
+            logger.info("The second execution of a sample produced no new edges. It will not be included in the corpus.")
+            return
+        }
+
+        var edges : [UInt32] = []
+        if origCov == secondCov {
+            edges = origCov.toEdges()
+        } else {
+            // Get the shared overlap between edges found
+            edges = origCov.edgeIntersection(otherEdgeSet: secondCov)
+            // Clear unshared edges
+            let edgeDifference = secondCov.edgeDifference(otherEdgeSet: origCov)
```

Comment:
IIUC, this computes the edges in `secondCov` that are not part of `origCov`. Don't you also need to clear the edges that are in `origCov` but not in `secondCov` (i.e. `origCov.edgeDifference(otherEdgeSet: secondCov)`)?

Reply:
All newly found edges in the first execution should be cleared here. https://github.com/googleprojectzero/fuzzilli/pull/171/files/79c0896fdcc6d49da4dbab8af446137af6c3e3af#diff-d11396d14389331fb850f0bc5f4d8ecee8aff27c291b9f114b4a47dda952ee6bR73

The second execution of a sample then cleanly obtains a list of edges, with the edges found as new in the second execution and shared with the first execution being counted as deterministic. Any edges found only by the second execution are then cleared in the linked code.

Reply:
Ah right, ok!